packages/lib-panoptes-js/src/experimental-auth.js

```diff
@@ -0,0 +1,238 @@
+    // - (If you're curious about the mechanics, check the first response headers'
+    //   "Set-Cookie" values, and the following second request header's "Cookie"
+    //   values.)
+    // - Our only control as front-end devs is to specify the fetch() of Request()'s
+    //   `credentials` option to either "omit" (bad idea for us), "same-origin"
+    //   (the default), or "include" (when you need things to work cross-origin)
+    // - ❗️ That `credentials: "include"` option is probably important if we need
+    //   Panoptes.JS to work on non-*.zooniverse.org domains!
+
+    const request2 = new Request(`https://panoptes-staging.zooniverse.org/users/sign_in`, {
+      body: JSON.stringify({
+        authenticity_token: csrfToken,
+        user: {
+          login,
+          password,
+          remember_me: true,
+        },
+      }),
+      credentials: 'include',
+      method: 'POST',
+      headers: PANOPTES_HEADERS,
+    })
+    const response2 = await fetch(request2)
+    console.log('+++ Step 2: login status: ', response2)
+
+    // Extract data and check for errors.
+    // NOTE: this is way more thorough than necessary.
+    if (!response2.ok) {
+      const jsonData2 = await response2.json()
+      const error = jsonData2?.error || 'No idea what went wrong; no specific error message detected.'
+      throw new Error(`Error from API. ${error}`)
+    }
+    const jsonData2 = await response2.json()
+    const userData = jsonData2?.users?.[0]
+    if (!userData) {
+      throw new Error('Impossible API response. No user returned.')
+    } else if (userData.login && userData.login !== login) {
+      throw new Error('Impossible API response. User returned is different from login attempt. Did you forget to sign out first?')
+    }
+
+    // Possible responses from /users/sign_in:
+    // 1. valid login & password => 200 with User resource 
+    // 2. blank login / password => 401 with { error: "You need to sign in or sign up before continuing." }
+    // 3. invalid login / password => 401 with { error: "Invalid email or password." }
+    // 4. invalid authenticity token => see https://github.com/zooniverse/operations/issues/561
+    // 5. already logged in => 200 with User resource of the previously logged-in user(!) This means if Panoptes thinks you're already logged in (see notes on http-only cookies), then any subsequent login attempts are ignored.
+    // X. Unexpected error, e.g. network down.
+
+    // Note: old PJC doesn't actually care about the response body, which is the
+    // logged-in user's User resource. This is because PJC has a separate call for
+    // fetching the user resource, which is, uh, kinda extra work, but keeps the
+    // calls standardised I guess?
+
+    // Step 3: get the bearer token.
+
+    const request3 = new Request(`https://panoptes-staging.zooniverse.org/oauth/token`, {
+      body: JSON.stringify({
+        client_id: '535759b966935c297be11913acee7a9ca17c025f9f15520e7504728e71110a27',
```

Comment:
Apps like Community Catalogue, which use the new API client, will need to pass in their own client ID here.

Reply:
Roger's auth client used a pattern like `const myAuthClient = new AuthClient({ clientId })`, which might work here too.